HT6_3.py

```diff
@@ -0,0 +1,228 @@
+
+    def get_due_date(self):
+        print('Type advertising expiration date')
+        while True:
+            exp_year = int(input('Enter a year: '))
+            if 2020 < exp_year <= 2024:
+                break
+            print('Year must be between 2021 and 2024')
+        while True:
+            exp_month = int(input('Enter a month: '))
+            if 1 <= exp_month <= 12:
+                break
+            print('Month must be a number between 1 and 12')
+        while True:
+            exp_day = int(input('Enter a day: '))
+            if 1 <= exp_day <= 31:
+                break
+            print('Day must be a number between 1 and 31')
+        exp_date = datetime.date(exp_year, exp_month, exp_day)
+        print('Expiration date is ', exp_date)
+        date_diff = datetime.date(exp_year, exp_month, exp_day) - datetime.date.today()
+        expiration_date = f'Actual until: {exp_date}, {date_diff.days} days left'
+        if date_diff.days < 0:
+            print('Your advertising is expired. It will not be published.')
+            exit()
+        return expiration_date
+
+
+class Joke(Publication):
+    def __init__(self, publication_type, publication_body:str ="", funny_meter: str = "") -> None:
+        super().__init__(publication_type=publication_type, publication_body=publication_body)
+        self.funny_meter = funny_meter
+
+    def generate_random_funny_meter(self):
+        value_list = ['One', 'Two', 'Three', 'Four', 'Five', 'Six', 'Seven', 'Eight', 'Nine', 'Ten']
+        return f'Funny meter – {value_list[random.randint(0, 9)]} of Ten'
+
+
+class Reader:
+    def __init__(self, publication_method):
+        self.publication_method = publication_method
+
+
+class ReaderCli(Reader):
+    def __init__(self):
+        super().__init__(publication_method="cli")
+
+    def __is_valid_publication_type(self, publication_type):
+        if publication_type not in  ["1", "2", "3"]:
+            print('Invalid record type')
+        return publication_type
+
+    def get_publication_type(self):
+        while True:
+            publication_type = input("What are you going to publish?\n1 - News\n2 - Ad\n3 - Joke\n")
+            if self.__is_valid_publication_type(publication_type):
+                if publication_type == "1":
+                    return News(publish_datetime=datetime.datetime.now(), publication_type="News")
+                elif publication_type == "2":
+                    return Ad(publication_type="Ad")
+                elif publication_type == "3":
+                    return Joke(publication_type="Joke")
+            else:
+                print("Wrong format. Please try again.")
+
+    def get_publication_text(self):
+        body = input('Type text: ')
+        return body
+
+    def get_publication_specific_info(self, publication: Publication) -> str:
+        if publication.publication_type == "News":
+            publication.location = publication.get_news_location()

```

Comment:
Actually, if you are referring to the functions of the child classes in the method, it is better to refer to typing.Union[News,Ad,Joke,OtherChildClass] in the signature